### PR TITLE
Restore client-side remote address in traces.

### PR DESCRIPTION
**Pull request by dsymonds**
The client-side traces were otherwise only showing `RPC: to <nil>`,
which is not helpful.

Also clean up construction of traceInfo and firstLine in a few places.

### CODE REVIEW

File: stream.go [hunk 1]
```diff
@@ -906,8 +914,8 @@ func (a *csAttempt) recvMsg(m interface{}, payInfo *payloadInfo) (err error) {
 
 func (a *csAttempt) finish(err error) {
 	a.mu.Lock()
+	defer a.mu.Unlock()
```
Comment (dfawley) [inline stream.go]: I'm not sure how much this affects things, but the reason we were not using `defer` here is for performance on the critical RPC path:

https://github.com/golang/go/issues/14939

Please revert these changes since they do not appear essential to the PR.

File: stream.go [hunk 2]
```diff
@@ -230,10 +230,14 @@ func newClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 	if c.creds != nil {
 		callHdr.Creds = c.creds
 	}
-	var trInfo traceInfo
+	var trInfo *traceInfo
```
Comment (dfawley) [inline stream.go]: If I'm not reading this wrong, this variable is only accessed inside the `if` below; then you can remove this line and use `:=` on line 235.

File: stream.go (hunk 1 above)
Comment (dsymonds) [reply stream.go]: It seems marginal at best given that this only happens at the end of a conn stream, and I don't see a benchmark verifying that it is noticeable, and the risk of forgetting it in an early return is significant, but your call; done.

File: stream.go (hunk 2 above)
Comment (dsymonds) [reply stream.go]: No, it's used further down in the call to `newAttemptLocked`.

Comment (dfawley) [reply stream.go]: Argh, Chrome ctrl+f fail with dynamically loaded content.

File: stream.go (hunk 1 above)
Comment (dfawley) [reply stream.go]: :man_shrugging: for one usage, you are right that it's minor, but it added up to something significant in our benchmarks since we had several other functions with defers on the RPC path.

Best to just avoid `defer` on any critical path until golang/go#14939 is fixed.
